firebase/extensions
# fix: add support for SendGrid categories and streamline SendGrid detection

This PR addresses the following improvements and fixes for the Firestore-Send-Email extension:

Enhancement: Increase in Payload Fields
Developers can now include categories.
Introduced a dynamic method for detecting SendGrid usage based on the SMTP_CONNECTION_URI. 
Bug Fix: Categories Not Being Sent
The categories field was not properly passed to the SendGrid API in the previous implementation.
Modified the email preparation and transport logic to ensure that the categories field is passed as per SendGrid's requirements.
Refactored Code and Tests
Removed the unused nodemailer-sendgrid dependency to simplify the codebase and avoid confusion.
Updated helpers.ts and corresponding tests in helpers.test.ts to support the new dynamic SendGrid detection.
Added logs to indicate whether the email is being sent via SendGrid or the standard transport.
Changes:

Code Changes:
Updated src/index.ts to dynamically detect SendGrid and utilize the appropriate sending method.
Modified src/helpers.ts to add an isSendGrid function and streamline SMTP transport setup.
Updated src/types.ts to reflect the additional flexibility in payload fields.
Tests:
Refactored helpers.test.ts to test the new isSendGrid function.
Added additional test cases to validate the proper handling of various SMTP connection URIs.
Dependencies:
Removed nodemailer-sendgrid from package.json and package-lock.json.
How to Test:

Deploy the updated extension in a local Firebase project with the emulators enabled.
Use the following payload in the Firestore mail collection:
```
{
  "to": ["example@example.com"],
  "categories": ["Example_Category"],
  "message": {
    "subject": "Final Test for Categories",
    "text": "This is a test email to see if categories work.",
    "html": "<strong>This is a test email to see if categories work.</strong>"
  }
}
```

Verify:
The email is successfully sent.
The categories field is correctly registered in SendGrid.
Logs indicate whether the SendGrid or standard transport was used.

Linked Issues:

Resolves #2053
Reviewer Checklist:
 Code adheres to the existing style and conventions.
 New features are covered by unit tests.
 Breaking changes are documented and communicated effectively.
